sixel.go

```diff
@@ -1,69 +1,69 @@
 package main
 
 import (
-	"fmt"
+	"log"
 	"os"
-	"strings"
+	"strconv"
 
 	"github.com/gdamore/tcell/v2"
 )
 
-const (
-	gSixelBegin = "\033P"
-
-	// The filler character should be:
-	// - rarely used: the filler is used to trick tcell into redrawing, if the
-	//   filler character appears in the user's preview, that cell might not
-	//   be cleaned up properly
-	// - ideally renders as empty space: the filler alternates between bold
-	//   and regular, using a non-space would look weird to the user.
-	gSixelFiller = '\u2000'
-)
+const gSixelBegin = "\033P"
 
 type sixelScreen struct {
-	xprev, yprev int
-	sixel        *string
-	altFill      bool
-	lastFile     string // TODO maybe use hash of sixels instead to flip altFill
+	lastFile   string
+	lastWin    win
+	forceClear bool
 }
 
-func (sxs *sixelScreen) fillerStyle(filePath string) tcell.Style {
-	if sxs.lastFile != filePath {
-		sxs.altFill = !sxs.altFill
+func (sxs *sixelScreen) clearSixel(win *win, screen tcell.Screen, filePath string) {
+	if sxs.lastFile != "" && (filePath != sxs.lastFile || *win != sxs.lastWin || sxs.forceClear) {
+		screen.LockRegion(sxs.lastWin.x, sxs.lastWin.y, sxs.lastWin.w, sxs.lastWin.h, false)
 	}
+}
 
-	if sxs.altFill {
-		return tcell.StyleDefault.Bold(true)
+func (sxs *sixelScreen) printSixel(win *win, screen tcell.Screen, reg *reg) {
+	if reg.path == sxs.lastFile && *win == sxs.lastWin && !sxs.forceClear {
+		return
 	}
-	return tcell.StyleDefault
-}
 
-func (sxs *sixelScreen) showSixels() {
-	if sxs.sixel == nil {
+	if reg.sixel == nil {
+		sxs.lastFile = ""
 		return
 	}
 
-	// XXX: workaround for bug where quitting lf might leave the terminal in bold
-	fmt.Fprint(os.Stderr, "\033[0m")
+	ti, err := tcell.LookupTerminfo(os.Getenv("TERM"))
+	if err != nil {
+		log.Printf("sixel: failed to look up term into %s", err)
+		return
+	}
 
-	fmt.Fprint(os.Stderr, "\0337")                              // Save cursor position
-	fmt.Fprintf(os.Stderr, "\033[%d;%dH", sxs.yprev, sxs.xprev) // Move cursor to position
-	fmt.Fprint(os.Stderr, *sxs.sixel)                           //
-	fmt.Fprint(os.Stderr, "\0338")                              // Restore cursor position
-}
+	tty, ok := screen.Tty()
+	if !ok {
+		log.Printf("sixel: failed to get tty")
+		return
+	}
 
-func (sxs *sixelScreen) printSixel(win *win, screen tcell.Screen, reg *reg) {
-	if reg.sixel == nil {
+	ws, err := tty.WindowSize()
+	if err != nil {
+		log.Printf("sixel: failed to get window size %s", err)
 		return
 	}
+	cw, ch := ws.CellDimensions()
```

Comment:
It's worth noting that these values can be 0, which causes a panic when you attempt to divide by them later. I assume tcell is getting them from the `TIOCGWINSZ` pixel dimensions which aren't guaranteed to be set.